pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -1751,6 +1751,103 @@ impl Timeline {
+            start: Key::MIN,
+            end: Key::MAX,
+        });
+        let compact_lsn = if let Some(compact_lsn) = options.compact_below_lsn {
+            compact_lsn
+        } else {
+            let gc_info = self.gc_info.read().unwrap();
+            gc_info.cutoffs.select_min() // use the real gc cutoff
+        };
+        let mut compact_jobs = Vec::new();
+        // For now, we simply use the key partitioning information; we should do a more fine-grained partitioning
+        // by estimating the amount of files read for a compaction job. We should also partition on LSN.
+        let partition = self.partitioning.lock().await;
+        let ((dense_ks, sparse_ks), _) = &*partition;
+        // Truncate the key range to be within user specified compaction range.
+        fn truncate_to(
+            source_start: &Key,
+            source_end: &Key,
+            target_start: &Key,
+            target_end: &Key,
+        ) -> Option<(Key, Key)> {
+            let start = source_start.max(target_start);
+            let end = source_end.min(target_end);
+            if start < end {
+                Some((*start, *end))
+            } else {
+                None
+            }
+        }
+        let mut split_key_ranges = Vec::new();
+        for partition in &dense_ks.parts {
+            for range in partition.ranges.iter() {
+                let Some((start, end)) = truncate_to(
+                    &range.start,
+                    &range.end,
+                    &compact_range.start,
+                    &compact_range.end,
+                ) else {
+                    continue;
+                };
+                split_key_ranges.push((start, end));
+            }
+        }
+        for partition in &sparse_ks.parts {
+            for range in partition.0.ranges.iter() {
+                let Some((start, end)) = truncate_to(
+                    &range.start,
+                    &range.end,
+                    &compact_range.start,
+                    &compact_range.end,
+                ) else {
+                    continue;
+                };
+                split_key_ranges.push((start, end));
+            }
+        }
+        let guard = self.layers.read().await;
+        let layer_map = guard.layer_map()?;
+        let mut current_start = None;
+        // Split compaction job to about 2GB each
+        const GC_COMPACT_MAX_SIZE_MB: u64 = 4 * 1024; // 2GB, TODO: should be configuration in the future
+        let ranges_num = split_key_ranges.len();
+        for (idx, (start, end)) in split_key_ranges.into_iter().enumerate() {
+            if current_start.is_none() {
+                current_start = Some(start);
+            }
+            let start = current_start.unwrap();
+            let res = layer_map.range_search(start..end, compact_lsn);
+            let total_size = res.found.keys().map(|x| x.layer.file_size()).sum::<u64>();
```

Comment:
Why are we not retrieving the `struct Layers` here and store them inside the `Jobs` struct?
Effectively what this function does is perform the layer selection.
Feels wrong to re-do the layer selection when we actually execute the scheduled job.

Reply:
The layers could change between each compaction runs. In the split function, we set the LSNs and split key ranges so that multiple partial compaction jobs cover the full range of a full compaction job. The actual layers to be compacted are determined at the time the compaction gets running.